File: results/dnd-rest-benchmark/lifecycle-runs/20260730T082348Z_claude_claude-sonnet-5_go-stdlib/play_migrations.go
package main

import (
	"encoding/json"
	"net/http"
)

// playMigration is the applied, deterministically upgraded state produced by
// migrating a compatible legacy schema version 1 snapshot, set only by the
// campaign DM.
type playMigration struct {
	SchemaVersion int    `json:"schema_version"`
	Story         string `json:"story"`
	CampaignName  string `json:"campaign_name"`
}

// handlePlayCampaignMigrationsSub routes the "migrations" and
// "migration-state" sub-paths of a play campaign.
func handlePlayCampaignMigrationsSub(w http.ResponseWriter, r *http.Request, id, rest string) bool {
	if rest == "migrations" {
		handleCreatePlayMigration(w, r, id)
		return true
	}

	if rest == "migration-state" {
		handleGetPlayMigrationState(w, r, id)
		return true
	}

	return false
}

func handleCreatePlayMigration(w http.ResponseWriter, r *http.Request, campaignID string) {
	if r.Method != http.MethodPost {
		writeError(w, http.StatusMethodNotAllowed, "method not allowed")
		return
	}

	username, ok := authenticatePlay(r)
	if !ok {
		writeError(w, http.StatusUnauthorized, "authentication required")
		return
	}

	var req struct {
		SchemaVersion int    `json:"schema_version"`
		Story         string `json:"story"`
	}
	if err := json.NewDecoder(r.Body).Decode(&req); err != nil {
		writeError(w, http.StatusBadRequest, "invalid request body")
		return
	}

	playMu.Lock()
	c, ok := requirePlayCampaign(w, campaignID)
	if !ok {
		return
	}
	if c.Owner != username {
		playMu.Unlock()
		writeError(w, http.StatusForbidden, "only the campaign dm may create a migration")
		return
	}

	if req.SchemaVersion != 1 || req.Story == "" {
		playMu.Unlock()
		writeError(w, http.StatusBadRequest, "invalid legacy schema snapshot")
		return
	}

	status := http.StatusCreated
	if c.Migration != nil && c.Migration.Story == req.Story {
		status = http.StatusOK
	} else {
		c.Migration = &playMigration{
			SchemaVersion: 2,
			Story:         req.Story,
			CampaignName:  c.Name,
		}
	}
	resp := map[string]interface{}{
		"schema_version": c.Migration.SchemaVersion,
		"story":          c.Migration.Story,
		"campaign_name":  c.Migration.CampaignName,
	}
	playMu.Unlock()
	if status == http.StatusCreated {
		persistState()
	}

	writeJSON(w, status, resp)
}

func handleGetPlayMigrationState(w http.ResponseWriter, r *http.Request, campaignID string) {
	if r.Method != http.MethodGet {
		writeError(w, http.StatusMethodNotAllowed, "method not allowed")
		return
	}

	username, ok := authenticatePlay(r)
	if !ok {
		writeError(w, http.StatusUnauthorized, "authentication required")
		return
	}

	playMu.Lock()
	c, ok := requirePlayCampaign(w, campaignID)
	if !ok {
		return
	}
	if c.Owner != username {
		playMu.Unlock()
		writeError(w, http.StatusForbidden, "only the campaign dm may read migrated state")
		return
	}

	if c.Migration == nil {
		playMu.Unlock()
		writeError(w, http.StatusNotFound, "no migration has been applied")
		return
	}

	resp := map[string]interface{}{
		"schema_version": c.Migration.SchemaVersion,
		"story":          c.Migration.Story,
		"campaign_name":  c.Migration.CampaignName,
	}
	playMu.Unlock()

	writeJSON(w, http.StatusOK, resp)
}

File: results/dnd-rest-benchmark/lifecycle-runs/20260730T075550Z_pi_kimi-k2p7-code_ruby-sinatra/lib/routes/play_migrations.rb
# frozen_string_literal: true

# Route definitions for play campaign schema migrations.

# --- Migrate a version-1 campaign snapshot to version 2 ---
post '/v1/play/campaigns/:id/migrations' do
  content_type :json
  campaign_id = params[:id]
  validate_campaign_id!(campaign_id)

  user = authenticate_actor!
  username = user[:username]
  campaign = load_play_campaign!(campaign_id)
  require_play_campaign_owner!(campaign, username)

  body = parse_json_body
  validate_migration_snapshot!(body)

  story = body['story']
  existing = Storage.load_play_campaign_migration(campaign_id)

  if existing && existing[:story] == story
    halt 200, JSON.dump(existing)
  end

  migrated = {
    schema_version: 2,
    story: story,
    campaign_name: campaign[:name]
  }
  Storage.save_play_campaign_migration(campaign_id, migrated[:schema_version], migrated[:story], migrated[:campaign_name])

  halt 201, JSON.dump(migrated)
end

# --- Read migrated version-2 campaign state ---
get '/v1/play/campaigns/:id/migration-state' do
  content_type :json
  campaign_id = params[:id]
  validate_campaign_id!(campaign_id)

  user = authenticate_actor!
  username = user[:username]
  campaign = load_play_campaign!(campaign_id)
  require_play_campaign_owner!(campaign, username)

  migrated = Storage.load_play_campaign_migration(campaign_id)
  json_error(404, 'migration state not found') unless migrated

  JSON.dump(migrated)
end
